### ISSUE TITLE
Documentation around query performance

### ISSUE THREAD

**Issue by imathews** [redivis/help#16]
See #14

**Comment by imathews** [comment]
It looks like this query finally completed after ~2hrs; there was no error here, it was just a very complex query. 2hrs is still exceptionally long, so I did a bit of a deeper dive to understand what was happening. Behind the scenes, Redivis leverages BigQuery — generally, it is incredibly fast, and shouldn’t really be affected by high user volumes, though there are obviously limits to its capabilities. https://cloud.google.com/bigquery/docs/best-practices-performance-compute

When looking into the query job statistics, it appears the vast majority of the query’s time was spent in computing all of these new variables — which makes sense when you consider the complexity of each statement (take a look at the code to view the full logical complexity). Removing the join, tweaking some other options really had no impact on the job’s performance. 

However, one action that did have a pronounced impact (cut run time in half) was removing the last CASE statement that referenced the other variables, and replacing that with a row filter (see screenshot). When creating a new variable that references other new variables, Redivis inlines those new variables directly into the code, as this is the only way to achieve this behavior via SQL. However, this doubles the compute complexity for our statement, assuming all of these CASE variables are kept — we’re basically computing each CASE statement twice.

I also wonder if these CASE expressions could be expressed as one large row filter. I recognize that this could get tricky to lay out, but I wouldn’t be surprised if this offers order-of-magnitude improvements over the case statements — it allows for lazy evaluation; once one of the conditions is true for a record, the compute engine no longer needs to solve for any of the subsequent conditions. I didn’t try this, but if you do, do let me know what you find out.

TLDR — be careful inlining New Variables within other New Variables in the same block, as this basically re-evalutes the New Variable again. Consider using row filters instead of case blocks. And queries with hundreds of logical statements may well take a while to run. Though BigQuery, and Redivis, are constantly being improved — know that this is the slowest it’ll ever be :)

**Comment by imathews** [comment]
https://help.redivis.com/for-researchers/concepts/optimizing-performance
